### PR TITLE
Fix patching for promise subclasses + strengthen native promise checks

### PR DESCRIPTION
**Pull request by matthewloring**

### CODE REVIEW

File: test/native-promises.tap.js [hunk 1]
```diff
@@ -2198,3 +2198,21 @@ if (typeof Promise.prototype.chain === 'function') {
     }
   });
 }
+
+test('subclasses', function(t) {
+  if (nodeVersion[0] < 6) {
+    // class syntax is not supported before node 6.
+    t.end();
```
Comment (Qard) [inline test/native-promises.tap.js]: A t.pass() with some explanation for the skip might be nice here.

File: test/promise-subclass.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+'use strict';
+
+module.exports = (tap) => {
+  return class SubclassedPromise extends Promise {
+    then(onSuccess, onReject) {
+      tap.type(onSuccess, 'function');
+      tap.type(onReject, 'undefined');
+      return Promise.prototype.then.call(this, onSuccess, onReject);
```
Comment (Qard) [inline test/promise-subclass.js]: Any reason not to just do super.then(...) here?

File: test/native-promises.tap.js (hunk 1 above)
Comment (matthewloring) [reply test/native-promises.tap.js]: Fixed.

File: test/promise-subclass.js (hunk 2 above)
Comment (matthewloring) [reply test/promise-subclass.js]: Not that I can think of, I've updated it.

**Comment by Qard** [comment]
0.12 through 5.x seem to fail on Travis. Not sure why. On mobile at the moment, so can't dig in right now.

**Comment by matthewloring** [comment]
Everything should be green now (except for the known failures on 7.x).

**Comment by hayes** [comment]
Would be cool to add tests that instanceof checks work as expected on promises returned by then, and maybe ensure that if the subclass defines a Symbol.species property that that is also respected. I don't think there are issues with either of these cases, but tests are cheap

**Comment by matthewloring** [comment]
@hayes I'm not sure which `instanceof` checks you are referring to. Are you asking for checks that the result of `Promise.then` has a natively implemented `toString`?

**Comment by hayes** [comment]
On mobile, so not sure if this will be exactly right, but basically was thinking something like:
```javascript
const p = new Promises class(resolve => resolve(123)).then(() => 456)

p instanceof PromiseSubclass
p instanceof OriginalPromise
p instanceof WrappedPromise
```

**Comment by hayes** [comment]
And if Promise.prototype[Symbol.species] is defined, some of those might be false.

**Comment by hayes** [comment]
To clarify, not a requirement to merge this change, and not testing this change specifically, but would be good for ensuring proper subclassing going forward. It would be easy to write a version of the wrapper around then that doesn't properly handle this

**Comment by matthewloring** [comment]
I've added the tests you've described to check the new ES6 class based promise wrappers.

**Comment by matthewloring** [comment]
@hayes Could you take a look at the newly added tests?

File: test/native-promises.tap.js [hunk 3]
```diff
@@ -2198,3 +2198,39 @@ if (typeof Promise.prototype.chain === 'function') {
...
+  }
+
+  // Promise subclasses do 2 asserts per promise.
+  t.plan(13);
+
+  var SubclassedPromise = require('./promise-subclass.js');
+  var StandardSubclassedPromise = SubclassedPromise(t, false);
+  var SubclassedPromiseCustomSpecies = SubclassedPromise(t, true);
+
+  var s = StandardSubclassedPromise.resolve(42).then(function(val) {
```
Comment (hayes) [inline test/native-promises.tap.js]: doesn't look like `s` is ever used

**Comment by hayes** [comment]
I think it's good to go, I can merge it, but I don't think I have publish access

**Comment by matthewloring** [comment]
@Qard Are you able to publish and update cls?

**Comment by matthewloring** [comment]
Ping, when do you think this could be published?

**Comment by Qard** [comment]
Published. Sorry for the delay.

**Pull request merged**